File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SetPropertiesInsteadOfMethods.cs
using System;
using System.Linq;
using System.Collections;
using System.Collections.Generic;

static class Program
{
    static void SortedSet()
    {
        var sortedSet = new SortedSet<int>();

        _ = sortedSet.Min(); // Noncompliant {{"Min" property of Set type should be used instead of the "Min()" extension method.}}
        //            ^^^
        _ = sortedSet?.Min(); // Noncompliant {{"Min" property of Set type should be used instead of the "Min()" extension method.}}
        //             ^^^
        _ = sortedSet.Max(); // Noncompliant {{"Max" property of Set type should be used instead of the "Max()" extension method.}}
        //            ^^^
        _ = sortedSet?.Max(); // Noncompliant {{"Max" property of Set type should be used instead of the "Max()" extension method.}}
        //             ^^^

        Func<SortedSet<int>, int> funcMin = x => x.Min(); // Noncompliant
        Func<SortedSet<int>, int> funcMax = x => x.Max(); // Noncompliant

        SortedSet<int> DoWork() => null;

        DoWork().Min(); // Noncompliant
        DoWork().Max(); // Noncompliant

        DoWork()?.Min(); // Noncompliant
        DoWork()?.Max(); // Noncompliant

        new SortedSet<int> { 42 }.Min(); // Noncompliant
        new SortedSet<int> { 42 }.Max(); // Noncompliant

        sortedSet.Min(x => 42f); // Compliant, predicate used
        sortedSet?.Max(x => x); // Compliant, predicate used
    }

    static void DerivesFromSortedSet()
    {
        var sortedSetDerived = new DerivesFromSetType<int>();

        (true ? sortedSetDerived : sortedSetDerived).Min(); // Noncompliant
        (true ? sortedSetDerived : sortedSetDerived).Max(); // Noncompliant

        (sortedSetDerived ?? sortedSetDerived).Min(); // Noncompliant
        (sortedSetDerived ?? sortedSetDerived).Max(); // Noncompliant

        (sortedSetDerived ?? (true ? sortedSetDerived : sortedSetDerived)).Min(); // Noncompliant
        (sortedSetDerived ?? (true ? sortedSetDerived : sortedSetDerived)).Max(); // Noncompliant

        sortedSetDerived.Fluent().Fluent().Fluent().Fluent().Min(); // Noncompliant
        sortedSetDerived.Fluent().Fluent().Fluent().Fluent()?.Max(); // Noncompliant
        sortedSetDerived.Fluent().Fluent().Fluent()?.Fluent().Min(); // Noncompliant
        sortedSetDerived.Fluent().Fluent().Fluent()?.Fluent()?.Max(); // Noncompliant
        sortedSetDerived.Fluent().Fluent()?.Fluent().Fluent().Min(); // Noncompliant
        sortedSetDerived.Fluent().Fluent()?.Fluent().Fluent()?.Max(); // Noncompliant
        sortedSetDerived.Fluent().Fluent()?.Fluent()?.Fluent().Min(); // Noncompliant
        sortedSetDerived.Fluent().Fluent()?.Fluent()?.Fluent()?.Max(); // Noncompliant
        sortedSetDerived.Fluent()?.Fluent().Fluent().Fluent().Min(); // Noncompliant
        sortedSetDerived.Fluent()?.Fluent().Fluent().Fluent()?.Max(); // Noncompliant
        sortedSetDerived.Fluent()?.Fluent().Fluent()?.Fluent().Min(); // Noncompliant
        sortedSetDerived.Fluent()?.Fluent().Fluent()?.Fluent()?.Max(); // Noncompliant
        sortedSetDerived.Fluent()?.Fluent()?.Fluent().Fluent().Min(); // Noncompliant
        sortedSetDerived.Fluent()?.Fluent()?.Fluent().Fluent()?.Max(); // Noncompliant
        sortedSetDerived.Fluent()?.Fluent()?.Fluent()?.Fluent().Min(); // Noncompliant
        sortedSetDerived.Fluent()?.Fluent()?.Fluent()?.Fluent()?.Max(); // Noncompliant
        //                                                       ^^^

        sortedSetDerived.Min(x => x == 42f); // Compliant, comparer used
        sortedSetDerived?.Max(x => x == 42); // Compliant, comparer used
    }

    static void TrueNegatives()
    {
        var set = new SortedSet<int>();
        _ = set.Min; // Compliant
        _ = set.Max; // Compliant

        T Min<T>() => default(T);
        T Max<T>() => default(T);

        Min<int>(); // Compliant
        Max<int>(); // Compliant

        var doesNotDerive = new DoesNotDeriveFromSetType<int>();
        doesNotDerive.Min(); // Compliant, does not derive from Set type
        doesNotDerive.Max(); // Compliant, does not derive from Set type

        var hidden = new DerivesFromSetTypeButHidesEnumerableMethods<int>();
        hidden.Min(); // Compliant, hides LINQ's Min extension
        hidden.Max(); // Compliant, hides LINQ's Max extension

        dynamic dynamicSet = new SortedSet<int>();
        dynamicSet.Min(); // Compliant, dynamic type
        dynamicSet.Max(); // Compliant, dynamic type
    }
}

class DerivesFromSetType<T> : SortedSet<T>
{
    public DerivesFromSetType<T> Fluent() => this;
}

class DerivesFromSetTypeButHidesEnumerableMethods<T> : SortedSet<T>
{
    public T Min() => default(T);
    public T Max() => default(T);
}

class DoesNotDeriveFromSetType<T> : IEnumerable<T>
{
    public IEnumerator<T> GetEnumerator() => null;
    IEnumerator IEnumerable.GetEnumerator() => null;
}

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SetPropertiesInsteadOfMethods.vb
Imports System
Imports System.Linq
Imports System.Collections
Imports System.Collections.Generic
Imports System.Collections.Immutable

Module Program
    Sub SortedSet()
        Dim sortedSet = New SortedSet(Of Integer)()

        Dim dummy = Enumerable.Min(sortedSet) ' Noncompliant {{"Min" property of Set type should be used instead of the "Min()" extension method.}}
        dummy = Enumerable.Max(sortedSet) ' Noncompliant {{"Max" property of Set type should be used instead of the "Max()" extension method.}}

        Dim funcMin As Func(Of SortedSet(Of Integer), Integer) = Function(x) Enumerable.Min(x) ' Noncompliant
        Dim funcMax As Func(Of SortedSet(Of Integer), Integer) = Function(x) Enumerable.Max(x) ' Noncompliant

        Dim doWork As Func(Of SortedSet(Of Integer)) = Function() Nothing

        dummy = Enumerable.Min(doWork()) ' Noncompliant
        dummy = Enumerable.Min(doWork.Invoke()) ' Noncompliant

        dummy = Enumerable.Min(New SortedSet(Of Integer)({42})) ' Noncompliant
        dummy = Enumerable.Max(New SortedSet(Of Integer)({42})) ' Noncompliant

        dummy = Enumerable.Min(sortedSet, Function(x) 42) ' Compliant, predicate used
        dummy = Enumerable.Max(sortedSet, Function(x) 42.0F) ' Compliant, predicate used
    End Sub

    Sub DerivesFromSortedSet()
        Dim sortedSetDerived = New DerivesFromSetType(Of Integer)()

        Dim ternary = IIf(True, sortedSetDerived, sortedSetDerived)
        Enumerable.Min(ternary) ' Compliant - IIf returns un-typed object
        Enumerable.Max(ternary) ' Compliant - IIf returns untyped object

        Enumerable.Min(sortedSetDerived.Fluent().Fluent().Fluent().Fluent()) ' Noncompliant
        Enumerable.Max(sortedSetDerived.Fluent().Fluent().Fluent()?.Fluent()) ' Noncompliant
        Enumerable.Min(sortedSetDerived.Fluent().Fluent()?.Fluent().Fluent()) ' Noncompliant
        Enumerable.Max(sortedSetDerived.Fluent().Fluent()?.Fluent()?.Fluent()) ' Noncompliant
        Enumerable.Min(sortedSetDerived.Fluent()?.Fluent().Fluent().Fluent()) ' Noncompliant
        Enumerable.Max(sortedSetDerived.Fluent()?.Fluent().Fluent()?.Fluent()) ' Noncompliant
        Enumerable.Min(sortedSetDerived.Fluent()?.Fluent()?.Fluent().Fluent()) ' Noncompliant
        Enumerable.Max(sortedSetDerived.Fluent()?.Fluent()?.Fluent()?.Fluent()) ' Noncompliant
        '          ^^^
        Enumerable.Min(sortedSetDerived, Function(x) x = 42) ' Compliant, predicate used
        Enumerable.Max(sortedSetDerived, Function(x) x = 42.0F) ' Compliant, predicate used
    End Sub

    Sub TrueNegatives()
        Dim sortedSet = New SortedSet(Of Integer)()
        Dim dummy = sortedSet.Min ' Compliant
        dummy = sortedSet.Max ' Compliant
        dummy = sortedSet.Min() ' Compliant
        dummy = sortedSet.Max() ' Compliant

        Dim doesNotDerive = New DoesNotDeriveFromSetType(Of Integer)()
        doesNotDerive.Min() ' Compliant, does not derive from Set type
        doesNotDerive.Max() ' Compliant, does not derive from Set type
        dummy = Enumerable.Min(doesNotDerive) ' Compliant, does not derive from Set type
        dummy = Enumerable.Max(doesNotDerive) ' Compliant, does not derive from Set type
    End Sub
End Module

Class DerivesFromSetType(Of T)
    Inherits SortedSet(Of T)

    Public Function Fluent() As DerivesFromSetType(Of T)
        Return Me
    End Function
End Class

Class DoesNotDeriveFromSetType(Of T)
    Implements IEnumerable(Of T)

    Public Function Min() As Integer
        Return 42
    End Function

    Public Function Max() As Integer
        Return 42
    End Function

    Public Function GetEnumerator() As IEnumerator(Of T) Implements IEnumerable(Of T).GetEnumerator
        Return Nothing
    End Function

    Private Function IEnumerable_GetEnumerator() As IEnumerator Implements IEnumerable.GetEnumerator
        Return Nothing
    End Function
End Class
